File: src/main/java/com/hazza/LeetCodeByJava/SegmentThree/No101SymmetricTree.java
package com.hazza.LeetCodeByJava.SegmentThree;

import com.hazza.LeetCodeByJava.datastructure.TreeNode;

import java.util.Stack;

/**
 * Created with IntelliJ IDEA.
 *
 * Description: https://leetcode.com/problems/symmetric-tree/description/
 * User: HazzaCheng
 * Contact: hazzacheng@gmail.com
 * Date: 17-11-22
 * Time: 6:14 PM
 */
public class No101SymmetricTree {

    // Solution 1
    public boolean isSymmetric(TreeNode root) {
        if (root == null) return true;
        return isSame(root.left, root.right);
    }

    private boolean isSame(TreeNode leftPart, TreeNode rightPart) {
        int nu = (leftPart == null ? 0 : 1) + (rightPart == null ? 0 : 1);
        if (nu == 1) return false;
        if (nu == 0) return true;

        return leftPart.val == rightPart.val &&
                isSame(leftPart.left, rightPart.right) &&
                isSame(leftPart.right, rightPart.left);
    }

    // Solution 2
    public boolean isSymmetric1(TreeNode root) {
        if(root==null)  return true;

        Stack<TreeNode> stack = new Stack<>();
        TreeNode left, right;
        if(root.left!=null){
            if(root.right==null) return false;
            stack.push(root.left);
            stack.push(root.right);
        } else if(root.right!=null) return false;

        while(!stack.empty()){
            if(stack.size()%2!=0) return false;
            right = stack.pop();
            left = stack.pop();
            if(right.val!=left.val) return false;

            if(left.left!=null){
                if(right.right==null) return false;
                stack.push(left.left);
                stack.push(right.right);
            } else if(right.right!=null) return false;


            if(left.right!=null){
                if(right.left==null) return false;
                stack.push(left.right);
                stack.push(right.left);
            } else if(right.left!=null) return false;
        }

        return true;
    }
}

File: src/main/python/leetcode_by_python/tree/No101_Symmetric_Tree.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019-09-08
# https://leetcode.com/problems/symmetric-tree/
from src.main.python.leetcode_by_python.data_structure.TreeNode import TreeNode


class Solution:
    # solution 1
    def isSymmetric1(self, root: TreeNode) -> bool:
        if not root:
            return True
        return self.isSame(root.left, root.right)

    def isSame(self, leftPart, rightPart):
        if (leftPart is None) ^ (rightPart is None):
            return False
        if not leftPart and not rightPart:
            return True

        return leftPart.val == rightPart.val and self.isSame(
            leftPart.left, rightPart.right) and self.isSame(
            leftPart.right, rightPart.left)

    # solution 2
    def isSymmetric2(self, root: TreeNode) -> bool:
        if not root:
            return True

        stack = []
        if root.left and root.right:
            stack.append(root.left)
            stack.append(root.right)
        elif (root.left is None) ^ (root.right is None):
            return False

        while stack:
            right, left = stack.pop(), stack.pop()
            if right.val != left.val:
                return False

            if left.left and right.right:
                stack.append(left.left)
                stack.append(right.right)
            elif (left.left is None) ^ (right.right is None):
                return False

            if left.right and right.left:
                stack.append(left.right)
                stack.append(right.left)
            elif (left.right is None) ^ (right.left is None):
                return False

        return True

